shell/platform/android/io/flutter/embedding/engine/systemchannels/RestorationChannel.java

```diff
@@ -0,0 +1,65 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.systemchannels;
+
+import android.support.annotation.NonNull;
+import io.flutter.Log;
+import io.flutter.embedding.engine.dart.DartExecutor;
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugin.common.StandardMethodCodec;
+
+/**
+ * System channel to exchange restoration data between framework and engine.
+ *
+ * <p>The engine can obtain the current restoration data from the framework via this channel to
+ * store it on disk and - when the app is relaunched - provide the stored data back to the framework
+ * to recreate the original state of the app.
+ */
+public class RestorationChannel {
+  private static final String TAG = "RestorationChannel";
+
+  public RestorationChannel(@NonNull DartExecutor dartExecutor) {
+    MethodChannel channel =
+        new MethodChannel(dartExecutor, "flutter/restoration", StandardMethodCodec.INSTANCE);
+    channel.setMethodCallHandler(handler);
+  }
+
+  private byte[] dataForFramework;
+  private byte[] dataFromFramework;
+
+  /** Obtain the most current restoration data that the framework has provided. */
+  public byte[] getRestorationDataFromFramework() {
+    return dataFromFramework;
+  }
+
+  /** Set the restoration data that will be sent to the framework when the framework requests it. */
+  public void setRestorationDataForFramework(byte[] data) {
+    dataForFramework = data;
```

Comment:
This disjoint handshake probably works but explicit contracts seem safer to maintain than implicit contracts. Can't the ui component just explicitly send it to the framework without an intermediate buffer? (remember the channels already have a buffer in case the framework wasn't ready to read during engine initialization but even that doesn't seem to be the case here)

Reply:
While channels have a buffer, the buffering happens on the Dart side (see dart:ui's channelBuffers [1]). At the point where Android provides us with the instance state (i.e. during onRestoreInstanceState or onCreate) the dart side may not be running yet, so all messages send at this point are still lost. Therefore, I opted for caching it myself on the engine side and just making the framework retrieve it whenever it is ready to receive the data.

[1] https://master-api.flutter.dev/flutter/dart-ui/channelBuffers.html